crates/nu-parser/src/lex.rs

```diff
@@ -447,14 +447,59 @@ pub fn lex_signature(
     special_tokens: &[u8],
     skip_comment: bool,
 ) -> (Vec<Token>, Option<ParseError>) {
-    lex_internal(
+    let mut state = LexState {
         input,
+        output: Vec::new(),
+        error: None,
         span_offset,
+    };
+    lex_internal(
+        &mut state,
         additional_whitespace,
         special_tokens,
         skip_comment,
         true,
-    )
+        None,
+    );
+    (state.output, state.error)
+}
+
+#[derive(Debug)]
+pub struct LexState<'a> {
+    pub input: &'a [u8],
+    pub output: Vec<Token>,
+    pub error: Option<ParseError>,
+    pub span_offset: usize,
+}
+
+// Lex until the output is max_tokens longer than before the call, or until the input is exhausted.
+// The behaviour here is non-obvious (maybe non-useful) iff your additional_whitespace doesn't include newline:
+// If you pass `output` in a state where the last token is an Eol, this might *remove* tokens.
```

Comment:
I am honestly not familiar enough with the lexing step we have to make an substantive comment to this, but I would suggest making the key parts here a doccomment for folks new to the lexer (like me :D)

Reply:
Agreed, good call :)